keps/sig-auth/266-kubelet-client-certificate-bootstrap-rotation/README.md

```diff
@@ -0,0 +1,156 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [x] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] KEP approvers have approved the KEP status as `implementable`
+- [x] Design details are appropriately documented
+- [x] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] Graduation criteria is in place
+- [x] "Implementation History" section is up-to-date for milestone
+- [x] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+- [x] User-facing documentation has been created at https://kubernetes.io/docs/tasks/tls/certificate-rotation/
+
+## Summary
+
+Currently, a kubelet has a certificate/key pair that authenticates the kubelet to the kube-apiserver.
+The certificate is supplied to the kubelet when it is first booted, via an out of cluster mechanism.
+This proposal covers a process for obtaining the initial cert/key pair and rotating it as expiration 
+of the certificate approaches.
+
+### Goals
+
+1. Enable exchange of a short-lived credential for a client certificate on a new node when starting a Kubelet for the first time.
+2. Enable renewal/rotation of a Kubelet certificate as it approaches expiration.
+
+## Proposal
+
+### During Kubelet Boot Sequence
+
+1. Look on disk for an existing cert/key pair managed by the certificate manager.
+1. If there is an existing cert/key pair, load them.
+1. If there is no existing cert/key pair, look for cert/key data specified in the kubelet config file:
+    - values encoded in the kubeconfig (CertData, KeyData, CAData)
+    - file references in the kubeconfig (CertFile, KeyFile, CAFile)
+1. If the certificate is a bootstrap certificate, use the certificate to:
+    - generate a key
+    - create a certificate signing request
+    - request a signed certificate from the API server
+    - wait for a response
+    - Store the new cert/key pairs in the Kubelet's certificate directory
+
+### As Expiration Approaches
+
+Rotating a Kubelet client certificate will work by generating a new private key,
+issuing a new Certificate Signing Request to the API Server, safely updating the
+cert/key pair on disk, begin using the new cert/key pair.
+
+1. Store the new cert/key pairs in the kubelet's certificate directory.
+   This will allow the kubelet to have a place for storing the multiple 
+   cert/key pairs that it might have available at any given moment 
+   (because of a rotation in progress).
+    - When cert/key files are specified in the kubeconfig, these will be used if
+      a newer, rotated, cert/key pair does not exist.
+2. There will be a kubelet configuration option (`--rotate-certificates`) which must be set to true
+   to enable certificate bootstrapping and rotation, and a feature gate (`RotateKubeletClientCertificate`)
+   which must be enabled prior to the feature's graduation to GA
```

Comment:
OpenShift has been running this gate on by default since 1.9 and the issues were mostly settled in the 1.11 timeframe (most of them dealt with failures, ensuring kubelet bootstrap and rotation handoff worked correctly, and failure modes where restarting kubelet while partial certs existed were incorrect).  Also things like ignoring expired certs from disk correctly.

Reply:
added to implementation history for reference